[cuDSS] Add a sample code for reading an input matrix market file for a simple cuDSS resolution

rbourgeois33 opened the pull request:
This PR implements the same new example as [this closed one](https://github.com/NVIDIA/CUDALibrarySamples/pull/264), but is aligned with the new contribution policy, namely commits are signed, and the target branch is `main`.

**Note 1:** I set the C++ standard to 17 in the CMakeLists.txt for this example

**Note 2:** I have not followed the update of cuDSS since the time I first wrote the example. Some hard coded rules I implemented  such as  _"algorithms 1 and 2 are only for non sym / non hermitian matrices"_ may no longer apply. I have not double checked yet

Summary of features, taken from [the previous PR ](https://github.com/NVIDIA/CUDALibrarySamples/pull/264):

 - Added timers to the different parts of the resolution
 - Compute the final error (on host), taking into account the view type
 - implemented a easy to use command line input
 ```bash
 Usage: ./binary <alg: 0|1|2|3> <mtype: general|symmetric|hermitian|spd|hpd> <mview: full|lower|upper> <matrix_filename> [vector_filename (optional)] 
For instance, to reproduce the simple example:
./binary 0 symmetric upper matrix.mtx rhs.mtx 
```
- Use a vector filled with 1 if no rhs is given
- Display a warnings if algs 1 or 2 are used with a mtype that is not general
- Display a warning if the mview is not 'full' while mtype is 'general'
- Throw a runtime error if the reader finds elements in the upper / lower part of the matrix and the mtype is lower / upper
- Throw a runtime error if the reader does not find elements in both upper / lower parts of the matrix and the mview is full.
- We accept unsorted entries, and will sort them with std::sort, as well as empty rows.
- The reader does a small mock allocation if no file is found so that subsequent free operations don't fail.
- Throws an error if nnz in the header is != to the # of elements in the mtx file
- Throws an error is read i,j are out of bound i.e <0, or >= n
- Display a warning for empty rows (maybe its a bit too much)
- add a verbose mode for infos and errors

The whole thing can be unit tested as I did locally [here](https://github.com/rbourgeois33/my-cudss-input-deck/blob/main/unit_test_matrix_reader.cpp), but I dont wan't to add tests to this Sample set repo.

Thanks again @kvoronin for figuring out the legal details. Let me know if I can do anything else from here. Happy to help.

Cheers

rbourgeois33 committed:
[cuDSS] add matrix market reader example

Signed-off-by: Remi Bourgeois <Remi.Bourgeois@cea.fr>

thoasm commented on cuDSS/simple_matrix_market/simple_matrix_market.cpp:
```diff
@@ -0,0 +1,470 @@
+        !x_values_h || !b_values_h)
+    {
+        printf("Error: host memory allocation failed\n");
+        return -1;
```
You are mixing `printf` and `std::cerr` for errors. According to the standard, that works fine, but I prefer code that uses either the C-style `printf` or the C++-style `std::cout` (instead of both).

Since all of our examples use `printf`, I suggest to change all of the `std::cerr` to `printf` or, if you want to print to `stderr`, you can use `fprintf(stderr, "Error: This is wrong\n");`

thoasm commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,197 @@
+
+template <typename _TYPE_>
+int matrix_reader(std::string filename, int& n, int& nnz, int** csr_offsets_h,
+                  int** csr_columns_h, _TYPE_** csr_values_h, cudssMatrixViewType_t mview, bool verbose=true, bool verbose_err = true)
```
In C and C++, names with an underscore, followed by an uppercase letter are reserved. I would suggest to name it `value_type` as that is more descriptive (and more in-line with our other examples).

thoasm commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,197 @@
+    int file_line_count = 0;
+
+    //Tuple for the mtw lines
+    std::vector<std::tuple<int, int, _TYPE_>> entries;
```
Please include the header `<tuple>`.

thoasm commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,197 @@
+        // ! Do a mock malloc anyway so that subsequent free does not fail !
+        *csr_offsets_h = (int*)malloc(1);
+        *csr_columns_h = (int*)malloc(1);
+        *csr_values_h = (_TYPE_*)malloc(1);
```
I would suggest to set them to `NULL` in order to make it clear that an error happened. For the subsequent free, you can add a simple check: `if (ptr) free(ptr);`

thoasm commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,197 @@
+
+    int row = 0;
+    while (std::getline(file, line)) {
+        if (line.empty() || line[0] == '%') continue;
```
You need to check the header and validate that it is a supported format.

thoasm commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,197 @@
+
+        if (!foundSize) {
+            int dummy0, dummy1;
+            lineData >> n >> dummy0 >> dummy1;
```
The `array` Matrix Market format only lists `m n` for the size information in contrast to the `coordinate` format.

Note that you should verify that only a single right hand side is supplied, since you don't support more:
```suggestion
            int num_rhs;
            lineData >> n >> num_rhs;
            if (num_rhs != 1) {
                // fprintf or std::cerr the error that only a single RHS is supported
                return EXIT_FAILURE;
            }
```

thoasm commented on cuDSS/simple_matrix_market/simple_matrix_market.cpp:
```diff
@@ -0,0 +1,470 @@
+    printf("--- Resolution over ! --- \n");
+
+    printf("Residual L2 error ||Ax - b|| = %e\n", residual);
+    bool passed = (residual < 1e-5);
```
nit: I would swap the print order (saying that it is over first, then list the total time, followed by computing the residual norm).
If you changed the print above to `Start solving`, this one should then read `Solving complete`.

thoasm reviewed (changes requested):
Thank you very much for taking the time and extending our samples! We really appreciate it.

I like your detailed error messages and your approach for error checking. This is only missing in the MatrixMarket header processing from what I can see.

We also need to look at the license headers in the files (I'm currently asking a colleague to see if my suggestion would work for us).
My remaining comments are only minor (mostly about style and consistency).

thoasm commented on cuDSS/simple_matrix_market/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
I discussed it with a colleague, and adding your name (see the suggested change) seems to be the way to go.

rbourgeois33 commented on cuDSS/simple_matrix_market/simple_matrix_market.cpp:
```diff
@@ -0,0 +1,470 @@
+#include <vector>
+
+#include <cmath>
+#include <vector>
```
removed one extra `#include <math.h>` too

thoasm commented on cuDSS/simple_matrix_market/matrix_market_reader.h:
```diff
@@ -0,0 +1,250 @@
+            int num_rhs;
+            lineData >> n >> num_rhs;
+            if (num_rhs != 1) {
+                fprintf(stderr, "Only one RHS is supported, %d found.", num_rhs);
```
nit: Missing the newline in the end:

```suggestion
                fprintf(stderr, "Only one RHS is supported, %d found.\n", num_rhs);
```

thoasm reviewed (commented):
Thank you for the fixes, they are great.

I found a couple of nits.
For the formatting, I can offer to run our normal formatting on your branch (we have a non-public clang-format file).

The formatting is still in the early stages, so there is a chance it might change in the future.
For the commit messages, yes, rebasing and merging all of them into a single commit would be great.

thoasm commented:
Thanks, now everything is resolved!

@rbourgeois33 can you temporarily add me as a developer in your branch? I applied the clang-format, but I'm not allowed to push to your branch.

thoasm commented:
Actually, it might be easier for you to apply this patch file: 
[clang-format.patch](https://github.com/user-attachments/files/23240568/clang-format.patch)

Simply run `git apply clang-format.patch` in the repository. That might be easier for you (not sure if you can allow others to contribute to your repository).

You can make this a separate commit that just states "Apply clang-format"

rbourgeois33 committed:
[cuDSS] Apply clang-format

Signed-off-by: Remi Bourgeois <Remi.Bourgeois@cea.fr>

kvoronin commented on cuDSS/simple_matrix_market/simple_matrix_market.cpp:
```diff
@@ -0,0 +1,425 @@
+    double residual = compute_residual_error(n, csr_offsets_h, csr_columns_h,
+                                             csr_values_h, x_values_h, b_values_h, mview);
+
+    printf("Residual L2 error ||Ax - b|| = %e\n", residual);
```
```suggestion
    printf("Residual L2 norm ||Ax - b|| = %e\n", residual);
```

rbourgeois33 replied:
True ! this is indeed not the error, but the residual.

kvoronin reviewed (changes requested):
The code looks good to me. The only remaining item is adding some comments and prints to steer potential users towards `simple_residual` sample for more practical error analysis.

rbourgeois33 committed:
[cuDSS] Steer users better

- Add a suggested first-usage command when sample is wrongly called.
- Guide users toward the sample_residual example.
- Reproduce the sample example: matrix is spd.

Signed-off-by: Remi Bourgeois <Remi.Bourgeois@cea.fr>

thoasm reviewed (approved):
LGTM!
